---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# EnrichKit

<!-- badges: start -->
<!-- badges: end -->

The goal of EnrichKit is to **perform over-representation** test of a given gene set pair (*SignificantGenes* and *TotolGenes*) based on hypergeometric distribution (**Fisher's exact test**). Gene sets could possily be non-preserved co-expression modules, differentially expressed gene (DEG), genes flagged by significant SNPs and etc..

Currently, **six pathway/annotation databases** are integrated:

- [Gene Ontology](http://ensemblgenomes.org/info/access/biomart)
- [KEGG](https://www.genome.jp/kegg/)
- [Interpro](http://ensemblgenomes.org/info/access/biomart)
- [MeSH](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C10&q=The+MeSH+translation+maintenance+system%3A+structure%2C+interface+design%2C+and+implementation.&btnG=)
- [Reactome](https://reactome.org/download-data)
- [Molecular Signatures](https://data.broadinstitute.org/gsea-msigdb/msigdb/release/)

Also, gene identifiers could be *Ensembl Gene ID*, *EntrezID* or *HGNC Gene Symbol*.

Note current release can only support **Bos Taurus**, other organism might be included in future release.

## Example

Soppose we have identified 2 out 5 DEG in each of the two lactations. 

```{r example}
require(EnrichKit,quietly = TRUE)
Sig_lac1 =   c("ENSBTAG00000012594","ENSBTAG00000049850")
Sig_lac2 =   c("ENSBTAG00000009188","ENSBTAG00000001258")
Tot_lac1 = c("ENSBTAG00000012594","ENSBTAG00000049850","ENSBTAG00000018278","ENSBTAG00000021997","ENSBTAG00000008482")
Tot_lac2 = c("ENSBTAG00000009188","ENSBTAG00000001258","ENSBTAG00000021819","ENSBTAG00000019404","ENSBTAG00000015212")

GeneInfo = convertNformatID(GeneSetNames=c("lactation1","lactation2"),
                            SigGene_list = list(Sig_lac1,Sig_lac2),
                            TotalGene_list = list(Tot_lac1,Tot_lac2),
                            IDtype = "ens")
GeneInfo
```

What is special about using `README.Rmd` instead of just `README.md`? You can include R chunks like so:

```{r cars}
summary(cars)
```

You'll still need to render `README.Rmd` regularly, to keep `README.md` up-to-date.

You can also embed plots, for example:

```{r pressure, echo = FALSE}
plot(pressure)
```

In that case, don't forget to commit and push the resulting figure files, so they display on GitHub!






## Installation

You can install the released version of EnrichKit from [CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("EnrichKit")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("liulihe954/EnrichKit")
```

